### PR TITLE
toptabbar-people-employees

### PR DESCRIPTION
**Pull request by victor-kotov-arcadia**
branch kick off with partial functionality from origin/feature/people-view-template-and-navigation branch for PR.

Notes:
- top tab bar is used at People screen/tab
- for a moment it shows same content for any tab (Department, Room, Company) - employees list is a subset of company's employees - current user's department employees.

Known issues:
- the current version of Avatar component causes a visual issue - photos have wrong (default) size instead of size provided by style
- UI freeze on data update (common for all data related UI across the app)

### CODE REVIEW

File: mobile/src/reducers/const.ts [hunk 1]
```diff
@@ -1 +1 @@
-export const apiUrl = 'https://arcadia-assistant-dev.arcadialab.ru/api'; //TODO: fix hardcode
\ No newline at end of file
+export const apiUrl = 'http://localhost:5000/api'; //TODO: fix hardcode
```
Comment (serjKim) [inline mobile/src/reducers/const.ts]: Revert

File: mobile/src/people/employees-list.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    };
+
+    constructor(props: EmployeesListProps) {
+        super(props);
+        this.state = {
+            employeesForMyDepartment: []
+        };
+    }
+
+    public componentWillReceiveProps(nextProps: Readonly<EmployeesListProps>) {
```
Comment (serjKim) [inline mobile/src/people/employees-list.tsx]: No quite. This is logic should be placed in reducer combining with LOAD_EMPLOYEES_FOR_DEPARTMENT action...

File: mobile/src/people/avatar.tsx [hunk 3]
```diff
@@ -54,12 +53,12 @@ function validateEncodedImage(data: string) {
...
-        super(props);
-        this.state = {};
-    }
+    public state: AvatarState = {
+        size: NaN,
+        borderRadius: NaN
+    };
 
-    public onLayout = (e: LayoutChangeEvent) => {
+    public onLayout = (e: any) => {
```
Comment (santee) [inline mobile/src/people/avatar.tsx]: ?

File: mobile/src/people/avatar.tsx [hunk 4]
```diff
@@ -3,21 +3,20 @@ import { defaultAvatar } from './avatar-default';
...
-        alignItems: 'center'
-    },
     outerFrame: {
+        justifyContent: 'center',
+        alignItems: 'center',
         // Following attributes are default
         borderColor: '#2FAFCC',
-        borderWidth: 2
+        borderWidth: 2,
+        height: 200,
```
Comment (santee) [inline mobile/src/people/avatar.tsx]: Regarding height and width again. If we use onLayout method then we have its size. So no height, width or borderRadius should be specified in these styles. There is no such thing as default value for it.
If those are not specified yet (nulls during render), just do not display anything. @serjKim @demkalkov  can anybody see any issues with that logic?

File: mobile/src/people/avatar.tsx [hunk 5]
```diff
@@ -54,12 +53,12 @@ function validateEncodedImage(data: string) {
 }
 
 export class Avatar extends Component<AvatarProps, AvatarState> {
-    constructor(props: AvatarProps) {
-        super(props);
-        this.state = {};
-    }
+    public state: AvatarState = {
+        size: NaN,
```
Comment (santee) [inline mobile/src/people/avatar.tsx]: why do you use NaN? Basically it's not different from using PI. nulls should be used to indicate that no value is assigned yet.

File: mobile/src/people/avatar.tsx [hunk 6]
```diff
@@ -3,21 +3,20 @@ import { defaultAvatar } from './avatar-default';
 import { View, Image, StyleSheet, ViewStyle, ImageStyle, LayoutChangeEvent } from 'react-native';
 
 const styles = StyleSheet.create({
-    container: {
```
Comment (santee) [inline mobile/src/people/avatar.tsx]: where and why container styles have gone?

File: mobile/src/people/employees-list-item.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import { TouchableHighlight, StyleSheet, Platform, Text, View, Image } from 'react-native';
+
+import { Avatar } from '../people/avatar';
+import { Employee } from '../reducers/organization/employee.model';
+import { employeesListItemStyles as styles } from './styles';
+
+interface EmployeesListItemProps {
+    employee: Employee;
+    id: string;
```
Comment (santee) [inline mobile/src/people/employees-list-item.tsx]: where is it used?

File: mobile/src/people/employees-list-item.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+export class EmployeesListItem extends React.Component<EmployeesListItemProps> {
+    public render() {
+        const employeeName = this.props.employee ? this.props.employee.name.replace(',', '') : 'Unknown';
```
Comment (santee) [inline mobile/src/people/employees-list-item.tsx]: please remove `replace`. I think we'll have to define it one place (Employee class, I'd say), we'll do it later.

File: mobile/src/people/avatar.tsx (hunk 4 above)
Comment (serjKim) [reply mobile/src/people/avatar.tsx]: I don't see the ability to override the size, if we pass custom size through the props.

File: mobile/src/people/avatar.tsx (hunk 3 above)
Comment (victor-kotov-arcadia) [reply mobile/src/people/avatar.tsx]: I'll revert it to newest version of the component - this one was used as a quick fix for the issue with incorrect size of photo (for some reason better coded and styled component drawing incorrect Image) - I'll turn it back and try to figure out why it's happening.

File: mobile/src/people/employees-list-item.tsx (hunk 7 above)
Comment (victor-kotov-arcadia) [reply mobile/src/people/employees-list-item.tsx]: you mean id prop? It supposed to be used for further development as param for an employee selection -> navigation to its profile and picking up details by id.

File: mobile/src/people/employees-list-item.tsx (hunk 8 above)
Comment (victor-kotov-arcadia) [reply mobile/src/people/employees-list-item.tsx]: ok

File: mobile/src/people/home-people-screen.tsx [hunk 9]
```diff
@@ -18,10 +19,13 @@ const navBar =  new TopNavBar('People');
 
 class HomePeopleScreenImpl extends React.Component<PeopleScreenProps> {
     public static navigationOptions = navBar.configurate();
-
+    
     public render() {
+        
         return (
-            <Text>PeopleScreen. There are { this.props.departments.length} departments</Text>
+            <View style={{ height: 500, flex: 1 }}>
```
Comment (santee) [inline mobile/src/people/home-people-screen.tsx]: is there a reason for height being hardcoded?

File: mobile/src/reducers/user/user-info.reducer.ts [hunk 10]
```diff
@@ -1,23 +1,35 @@
...
 };
 
 export const userInfoReducer: Reducer<UserInfoState> = (state = initState, action: UserActions) => {
     switch (action.type) {
         case 'LOAD-USER-EMPLOYEE-FINISHED':
             return {
+                ...state,
                 employee: action.employee
             };
+        case 'LOAD-USER-DEPARTMENT-EMPOYEES-FINISHED':
```
Comment (serjKim) [inline mobile/src/reducers/user/user-info.reducer.ts]: You should add it to your reducer in People section.

**Comment by serjKim** [comment]
Conflicts

**Comment by victor-kotov-arcadia** [comment]
merging...

File: mobile/src/people/employees-list.tsx [hunk 11]
```diff
@@ -19,15 +20,19 @@ const mapStateToProps = (state: AppState): EmployeesListProps => ({
...
-        return (
+        const employees = this.props.employeesForMyDepartment;
+
+        return employees.size > 0 ? 
             <View style={styles.view}>
                 <FlatList
-                    data={this.props.employeesForMyDepartment.toArray()}
+                    data={employees.toArray()}
                     keyExtractor={this.keyExtractor}
                     renderItem={({ item }) => <EmployeesListItem id={item.employeeId} employee={item} />} />
```
Comment (serjKim) [inline mobile/src/people/employees-list.tsx]: Move to the separate method

File: mobile/src/people/employees-list-item.tsx [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        const employeePosition = this.props.employee ? this.props.employee.position : 'Unknown';
+        const photo = this.props.employee ? this.props.employee.photo : null;
+        const mimeType = photo ? photo.mimeType : null;
+        const base64 = photo ? photo.base64 : null;
+
+        return (
+            <TouchableHighlight>
+                <View style={styles.layout}>
+                    <View style={styles.avatarContainer}>
+                        <Avatar mimeType={mimeType} photoBase64={base64} style={StyleSheet.flatten(styles.avatar)} />
```
Comment (santee) [inline mobile/src/people/employees-list-item.tsx]: does it even compile? Avatar takes photo now.

File: mobile/src/reducers/people/people.action.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import { User } from '../user/user.model';
+import { Employee } from '../organization/employee.model';
+import { EmployeeMap } from '../organization/employees.reducer';
+
+export interface LoadUserDepartmentEmployessFinished {
+    type: 'LOAD-USER-DEPARTMENT-EMPOYEES-FINISHED';
```
Comment (santee) [inline mobile/src/reducers/people/people.action.ts]: typo

File: mobile/src/reducers/people/people.reducer.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+const initState: PeopleState = {
+    employees: Map()
+};
+
+export const peopleEpics = combineEpics(
+    loadUserDepartmentEmployeesEpic$ as any);
+
+export const peopleReducer: Reducer<PeopleState> = (state = initState, action: PeopleActions) => {
```
Comment (santee) [inline mobile/src/reducers/people/people.reducer.ts]: we have to discuss it tomorrow, I'm not sure what's the point of the separate reducer. Did @serjKim say to do it that way? :)

Comment (serjKim) [reply mobile/src/reducers/people/people.reducer.ts]: yes, initially

File: mobile/src/people/avatar.tsx [hunk 15]
```diff
@@ -12,18 +12,23 @@ const styles = StyleSheet.create({
...
         borderColor: '#2FAFCC',
-        borderWidth: 2
     },
     image: {
         borderColor: '#fff',
-        borderWidth: 2,
         flex: 1
+    },
+    default: {
+        height: 200,
```
Comment (santee) [inline mobile/src/people/avatar.tsx]: why height and width have default values? They are supposed to be taken from the actual size. BorderWidth / inner border only things that can be overriden

Comment (victor-kotov-arcadia) [reply mobile/src/people/avatar.tsx]: we'd also discussed it previously, these defaults for cases if no size was specified.

File: mobile/src/people/employees-list-item.tsx (hunk 12 above)
Comment (victor-kotov-arcadia) [reply mobile/src/people/employees-list-item.tsx]: compile but no real photo, just stub, I was wondering why - will be fixed :)

**Pull request closed without merging**